components/core/Application.js

```diff
@@ -424,6 +471,99 @@ export default class ApplicationPage extends React.Component {
     this._handleRegisterLoadingFinished({ keys });
   };
 
+  _handleOptimisticUpload = async ({ files, slate }) => {
+    let optimisticFiles = [];
+
+    for (let i = 0; i < files.length; i++) {
+      let id = uuid();
+      let dataURL = await this._handleLoadDataURL(files[i]);
+      let data = {
+        id,
+        filename: files[i].name,
+        data: {
+          name: files[i].name,
+          type: files[i].type,
+          size: files[i].size,
+        },
+        decorator: "OPTIMISTIC-FILE",
+        dataURL,
+      };
+
+      optimisticFiles.push(data);
+
+      files[i].id = id;
+    }
+
+    this.setState({ optimisticFiles });
+
+    if (slate && slate.id) {
+      const slates = this.state.viewer.slates;
+
+      for (let item of slates) {
+        if (item.id === slate.id) {
+          item.objects = [...item.objects, ...optimisticFiles];
+          break;
+        }
+      }
+
+      this._handleUpdateViewer({ slates });
+      return files;
+    }
+
+    let update = [...optimisticFiles, ...this.state.viewer?.library];
+    let library = this.props.viewer?.library;
+    library = update;
+    this._handleUpdateViewer({ library });
+
+    return files;
+  };
+
+  _handleLoadDataURL = (file) =>
+    new Promise((resolve, reject) => {
+      if (file.type.startsWith("application/pdf")) {
+        resolve(URL.createObjectURL(file));
+        return;
+      }
+
+      const reader = new FileReader();
+      reader.onload = () => {
+        resolve(reader.result);
+      };
+
+      reader.onerror = () => {
+        reject({ error: true });
+        reader.abort();
+      };
+
+      reader.readAsDataURL(file);
+    });
+
+  _handleSuccessfulUpload = ({ succeeded }) => {
+    let optimisticFiles = this.state.optimisticFiles;
+    // let optimisticFilesNames = this.state.optimisticFiles.map(item => item.name);
+    let library = this.state.viewer.library;
+    let update = succeeded.map((item) => {
+      let data = item.json?.data;
+
+      let itemToUpdateIndex = library[0].children.findIndex((item) => item.id === data.id);
```

Comment:
This is something that changed in the migration. Pleas make sure you take a look at the doc https://www.notion.so/slatesystem/Files-Database-Migration-84ccad66574d487fbad4a4a2a5857276 for changes.

It's no longer library[0].children, it's now just library

Reply:
This might be the source of some of the other errors I mentioned (like uploads not resolving properly etc)